File: yuv_convert/src/main/java/com/zmy/yuv_convert/input/ByteBufferInput.java
package com.zmy.yuv_convert.input;

import com.zmy.yuv_convert.Format;

import java.nio.ByteBuffer;

public class ByteBufferInput extends Input<ByteBuffer[]> {
    private ByteBuffer buffer;

    @Override
    public void provide(ByteBuffer[] data, Format format, int width, int height, int[] stride) {
        int totalLength = 0;
        for (ByteBuffer datum : data) {
            totalLength += datum.capacity();

        }
        int totalCapacity = Math.max(totalLength, format.getMinFrameSize(width, height));
        if (buffer == null || buffer.capacity() < totalCapacity) {
            buffer = ByteBuffer.allocateDirect(totalCapacity);
        }
        int start = 0;
        for (int component = 0; component < data.length; component++) {
            ByteBuffer it = data[component];
            it.position(0);
            it.limit(it.capacity());
            buffer.position(start);
            buffer.limit(start + it.capacity());
            buffer.put(it);
            start += Math.max(it.capacity(), format.getComponentCapacity(component, width, height));
        }
        setInputData(new InputData(buffer, format, width, height, stride));
    }
}

File: yuv_convert/src/main/java/com/zmy/yuv_convert/input/ByteBufferInput.kt
package com.zmy.yuv_convert.input

import com.zmy.yuv_convert.Format
import java.nio.ByteBuffer
import kotlin.math.max

class ByteBufferInput : Input<Array<ByteBuffer>>() {
    private var buffer: ByteBuffer? = null
    override fun provide(data: Array<ByteBuffer>, format: Format, width: Int, height: Int, stride: IntArray) {
        val totalCapacity = max(data.map { it.capacity() }.reduce { s, t -> s + t }, format.getMinFrameSize(width, height))
        if ((buffer?.capacity() ?: -1) < totalCapacity) {
            buffer = ByteBuffer.allocateDirect(totalCapacity)
        }
        var start = 0
        data.forEachIndexed { component, it ->
            it.position(0)
            it.limit(it.capacity())
            buffer?.position(start)
            buffer?.limit(start + it.capacity())
            buffer?.put(it)
            buffer?.position(0)
            start +=  max(it.capacity(), format.getComponentCapacity(component, width, height))
        }
        inputData = InputData(buffer!!, format, width, height, stride)
    }
}